platform/linux-generic/arch/aarch64/odp_crypto_armv8.c

```diff
@@ -137,6 +154,280 @@ null_crypto_routine(odp_packet_t pkt ODP_UNUSED,
 	return ODP_CRYPTO_ALG_ERR_NONE;
 }
 
+static
+odp_crypto_alg_err_t aes_gcm_encrypt(odp_packet_t pkt,
+				     const odp_crypto_packet_op_param_t *param,
+				     odp_crypto_generic_session_t *session)
+{
+	armv8_cipher_state_t cs = {
+		.counter = {
+			.d = {0, 0}
+		}
+	};
+	armv8_cipher_constants_t cc = {
+		.mode = 0
+	};
+	uint8_t *iv_ptr;
+	uint64_t nonce_length, nonce_byte_length;
+	uint64_t plaintext_length, plaintext_byte_length;
+	uint64_t aad_length, aad_byte_length;
+	int ret = 0;
+
+	/* Clear the allocated buffers */
+	memset(global->input, 0, _ARM_CRYPTO_MAX_DATA_LENGTH);
+	memset(global->nonce, 0, _ARM_CRYPTO_MAX_IV_LENGTH);
+	memset(global->aad, 0, _ARM_CRYPTO_MAX_AAD_LENGTH);
+	memset(global->digest, 0, _ARM_CRYPTO_MAX_DIGEST_LENGTH);
+	memset(global->output, 0, (_ARM_CRYPTO_MAX_DATA_LENGTH +
+	       _ARM_CRYPTO_MAX_DIGEST_LENGTH));
+
+	nonce_length = (session->p.cipher_iv.length) * 8;
+	/* IV length must be a multiple of 16 */
+	nonce_byte_length = ((nonce_length + 127) & ~127UL) >> 3;
+
+	if (param->cipher_iv_ptr)
+		iv_ptr = param->cipher_iv_ptr;
+	else if (session->p.cipher_iv.data)
+		iv_ptr = session->cipher.iv_data;
+	else
+		return ODP_CRYPTO_ALG_ERR_IV_INVALID;
+
+	/* Copy nonce and pad it to nonce_byte_length */
+	for (uint64_t i = 0; i < session->p.cipher_iv.length; i++)
+		global->nonce[i] = iv_ptr[i];
+	for (uint64_t i = session->p.cipher_iv.length; i < nonce_byte_length; i++)
+		global->nonce[i] = 0xde;
```

Comment:
What is 0xde?

The IV should be padded with all zeroes except the last bit should be one. But the IV padding is done by the library so this piece of code does not seem to serve any purpose.

Reply:
In the Armv8 crypto library, padding is not handled internally by the API and we must explicitly pass the padded variables (IV, AAD and plaintext) to the encryption function. This is demonstrated in the functional test that was written by the Armv8 crypto library maintainers: https://github.com/ARM-software/AArch64cryptolib/blob/master/test/aesgcm_test_functional.c Also, in this test, they use 0xde as the padding bytes which I have replicated in this code

Reply:
> In the Armv8 crypto library, padding is not handled internally by the API and we must explicitly pass the padded variables (IV, AAD and plaintext) to the encryption/decryption function. 

No, I do not think so, as I explained, if it is ok to do read access to less than a block size of memory after the input data. For output data (plaintext in decryption, ciphertext in encryption) requires write access to those additional bytes too, which is a bit nastier requirement, but not about padding the input data but ensuring the output buffer is large enough.

Actually, it seems to me that you could avoid the costly packet data copying in most cases by just giving pointer to the contiguous (not all packets are) packet data as input and similar packet data pointer (with large enough contiguous data area taking into account the overwrite requirement) as the output pointer. The copying is needed only if the packet data under the crypto operation is not contiguous in memory (segmented packets, but even then not necessarily if the crypto range falls within single segment) or if the output data just would not fit in a segment (the overwrite bytes extending beyond end of the writable area).

> Also, in this test, they use 0xde as the padding bytes which I have replicated in this code

I did not spend much time looking at the test code but the 0xde there does not really seem to me to be actual padding required by the library but test related addition to better detect bugs in the library. The mere fact that there is such 0xde there does not mean that it needs to be here, especially if it it seems unnecessary and there is no convincing explanation why it is not.

Reply:
I am removing the explicit padding for the IV, AAD and plaintext in the next version. Handling multiple segments is the only case scenario when packet data copying will be valid so I will make those changes separately.